### PR TITLE
Add cudagraph_dynamic_shape_rerecord_limit (#197433)

### PR DESCRIPTION
**Pull request by Microve**
Summary:

Currently we only have a warning when the number of re-recording is larger than a threshold. This is no sufficient, because it's common that some graphs have limited dynamic shape variants while some others are highly dynamic, for the first one, we can afford the rerecord and want to enable cudagraph there. But we don't have the control, we only have a global switch to turn on skip dynamic. By adding this config, we can specify the threshold we allow for re-recordings for dynamic shapes, so that limited dynamic graphs can have cudagraph while highly dynamic one can be skipped.

Test Plan:
**Without this change**
f1145748212
it keeps re-recording on graph 22
 {F1996665990}

**With this change**
f1145507968
It ran as expected, 5 for forward, 5 for backward
 {F1996665946}

Differential Revision: D120550840

### CODE REVIEW

File: torch/_inductor/cudagraph_trees.py [hunk 1]
```diff
@@ -467,7 +467,23 @@ def deferred_cudagraphify(inputs: list[InputType]) -> OutputType:
...
+            # graph rather than on every call that falls back to eager.
+            torch._logging.warning_once(
+                log,
+                "[%s] Skipping cudagraph re-recording for new dynamic shapes: hit the "
+                "cudagraph_dynamic_shape_rerecord_limit of %s distinct recorded "
+                "shapes. Already-recorded shapes still replay; new shapes run eager.",
+                compile_id,
+                limit,
+            )
+            return model(inputs)
```
Comment (eellison) [inline torch/_inductor/cudagraph_trees.py]: [TODO: my note]

> This fallback bypasses the tree manager. The forward and backward each have their own `cudagraphify_impl`/`fn_cache` and hit the limit independently, so a cudagraphed forward can be followed by an eager backward. In that case nothing calls `manager.set_to_running_backward()` (compare `maybe_handle_backward_generation` in output_code.py), so `running_forwards_with_pending_backwards` stays `True` and new generations can't start.
>
> Repro: shape A forward-only, then shapes B and C forward+backward, then A forward+backward in a loop while holding the previous loss alive. With `limit=2`, the forward records a new cudagraph every iteration (11 graphs and growing; 6 stable with no limit). This fixes it (verified: 4 stable graphs):
>
> ```python
> if kwargs.get("is_backward"):
>     manager = get_manager(kwargs["device_index"], create_if_none_exists=False)
>     if manager is not None:
>         manager.set_to_running_backward()
> return model(inputs)
> ```
>
> Please also add a training test (forward+backward, with the forward and backward caches filling in different orders).

File: torch/_inductor/cudagraph_trees.py [hunk 2]
```diff
@@ -467,7 +467,23 @@ def deferred_cudagraphify(inputs: list[InputType]) -> OutputType:
         fn = fn_cache.get(int_key)
         if fn is not None:
             return fn(inputs)
+
         compile_id = kwargs.get("compile_id", "")
+
+        limit = config.triton.cudagraph_dynamic_shape_rerecord_limit
```
Comment (eellison) [inline torch/_inductor/cudagraph_trees.py]: [TODO: my note]

> Two things about reading the config here:
> - It's read at call time, but `torch.compile(options=...)` / `config_patches` only patch config during compilation, and `cudagraphify_impl` itself is built lazily on the first call. So passing the limit through `options=` silently does nothing (verified: 4 recordings with `limit=2`). `cudagraph_managed_input_rerecord_limit` reads its config when the wrapper is built in `compile_fx.cudagraphify` and passes it as a kwarg; please do the same.
> - The check runs before the `int_key is None` case, so `limit=0` also turns off cudagraphs for graphs with no symints (verified). Either skip the check when `int_key is None` or require `limit >= 1`.

File: torch/_inductor/cudagraph_trees.py [hunk 3]
```diff
@@ -467,7 +467,23 @@ def deferred_cudagraphify(inputs: list[InputType]) -> OutputType:
...
+        if limit is not None and len(fn_cache) >= limit:
+            # warning_once caches on the args, so this fires once per compiled
+            # graph rather than on every call that falls back to eager.
```
Comment (eellison) [inline torch/_inductor/cudagraph_trees.py]: [TODO: my note]

> `warning_once` caches on `(log, msg, compile_id, limit)`. The forward, the backward and all partitions share one `compile_id`, so this fires once for all of them, not once per compiled graph. In my repro the backward's skip was silent. Consider adding `is_backward` to the message args, and fix the comment.

File: test/inductor/test_cudagraph_trees.py [hunk 4]
```diff
@@ -4867,6 +4867,54 @@ def forward(self, x) -> torch.Tensor:
...
+                    # Shapes past the limit run eager, and must still be correct.
+                    self.assertEqual(out, mod(inp))
+
+                # An already-recorded shape keeps replaying its graph rather than
+                # falling back to eager or re-recording.
+                for _ in range(3):
+                    compiled(inps[0])
+
+            # Only the first two distinct shapes recorded a cudagraph.
+            self.assertEqual(self.get_manager().new_graph_id().id, 2)
```
Comment (eellison) [inline test/inductor/test_cudagraph_trees.py]: [TODO: my note]

> This fails locally with `Expected 2 but got 4`: see the top-level comment. Using `mod.requires_grad_(False)` (or `torch.no_grad()`) gives 2, but then the test only covers inference; please add a separate training test for the backward path.

File: torch/_inductor/cudagraph_trees.py [hunk 5]
```diff
@@ -467,7 +467,23 @@ def deferred_cudagraphify(inputs: list[InputType]) -> OutputType:
         fn = fn_cache.get(int_key)
         if fn is not None:
             return fn(inputs)
+
         compile_id = kwargs.get("compile_id", "")
+
+        limit = config.triton.cudagraph_dynamic_shape_rerecord_limit
+        if limit is not None and len(fn_cache) >= limit:
```
Comment (eellison) [inline torch/_inductor/cudagraph_trees.py]: [TODO: my note]

> The limit is counted per `cudagraphify_impl` closure, which means per fwd/bwd graph *and* per graph partition, each with its own `fn_cache` over its own subset of symints. From reading the code (I haven't reproduced the partition case), partitions can hit the limit at different times, so within one call some partitions replay and others run eager, and total recordings can reach roughly (#partitions x limit).
>
> It would be more predictable to count once per (compilation, fwd/bwd): for example, create one budget per `CompiledFxGraph` in `cudagraph_post_compile` / `cudagraph_partition_post_compile` and pass it through the cudagraphify kwargs, so all partitions of the forward share it (and likewise for the backward).
>
> Also, `limit=0` disables cudagraphs even for graphs without symints (verified); please require `limit >= 1`.

**Review by eellison — APPROVED** [review]
Looks good - couple small nits on setting backward when invoked, and making sure we have the right cache here.
